core/src/main/java/org/apache/druid/data/input/impl/JsonReader.java

```diff
@@ -49,41 +73,57 @@
       boolean keepNullColumns
   )
   {
-    super(inputRowSchema, source);
+    this.inputRowSchema = inputRowSchema;
+    this.source = source;
     this.flattener = ObjectFlatteners.create(flattenSpec, new JSONFlattenerMaker(keepNullColumns));
     this.mapper = mapper;
   }
 
   @Override
-  public List<InputRow> parseInputRows(String line) throws IOException, ParseException
+  protected CloseableIterator<String> intermediateRowIterator() throws IOException
   {
-    final JsonNode document = mapper.readValue(line, JsonNode.class);
-    final Map<String, Object> flattened = flattener.flatten(document);
-    return Collections.singletonList(MapInputRowParser.parse(getInputRowSchema(), flattened));
+    return CloseableIterators.withEmptyBaggage(
+        Iterators.singletonIterator(IOUtils.toString(source.open(), StringUtils.UTF8_STRING))
+    );
   }
 
   @Override
-  public Map<String, Object> toMap(String intermediateRow) throws IOException
+  protected List<InputRow> parseInputRows(String intermediateRow) throws IOException, ParseException
   {
-    //noinspection unchecked
-    return mapper.readValue(intermediateRow, Map.class);
-  }
+    try (JsonParser parser = new JsonFactory().createParser(intermediateRow)) {
+      final Iterator<JsonNode> delegate = mapper.readValues(parser, JsonNode.class);
+      return FluentIterable.from(() -> delegate)
+                           .transform(jsonNode -> MapInputRowParser.parse(inputRowSchema, flattener.flatten(jsonNode)))
+                           .toList();
+    }
+    catch (RuntimeException e) {
+      //convert Jackson's JsonParseException into druid's exception for further processing
+      if (e.getCause() instanceof JsonParseException) {
```

Comment:
Where is `JsonParseException` thrown wrapped in `RuntimeException`? Can you add a comment about it?